src/random.cpp

```diff
@@ -278,118 +282,252 @@ void GetOSRand(unsigned char *ent32)
+{
+    unsigned char buffer[32];
+
+    // Stack pointer to indirectly commit to thread/callstack
+    const unsigned char* ptr = buffer;
+    hasher.Write((const unsigned char*)&ptr, sizeof(ptr));
 
-    memory_cleanse(&nPerfCounter1, sizeof(nPerfCounter1));
-    memory_cleanse(&nPerfCounter2, sizeof(nPerfCounter2));
+    // Hardware randomness is very fast when available; use it always.
+    bool have_hw_rand = GetHardwareRand(buffer);
+    if (have_hw_rand) hasher.Write(buffer, sizeof(buffer));
+
+    // High-precision timestamp
+    SeedTimestamp(hasher);
 }
 
+static void SeedSlow(CSHA512& hasher) noexcept
+{
+    unsigned char buffer[32];
 
-static Mutex cs_rng_state;
-static unsigned char rng_state[32] = {0};
-static uint64_t rng_counter = 0;
+    // Everything that the 'fast' seeder includes
+    SeedFast(hasher);
 
-static void AddDataToRng(void* data, size_t len) {
-    CSHA512 hasher;
-    hasher.Write((const unsigned char*)&len, sizeof(len));
-    hasher.Write((const unsigned char*)data, len);
-    unsigned char buf[64];
-    {
-        WAIT_LOCK(cs_rng_state, lock);
-        hasher.Write(rng_state, sizeof(rng_state));
-        hasher.Write((const unsigned char*)&rng_counter, sizeof(rng_counter));
-        ++rng_counter;
-        hasher.Finalize(buf);
-        memcpy(rng_state, buf + 32, 32);
-    }
-    memory_cleanse(buf, 64);
+    // OS randomness
+    GetOSRand(buffer);
+    hasher.Write(buffer, sizeof(buffer));
+
+    // OpenSSL RNG (for now)
+    RAND_bytes(buffer, sizeof(buffer));
+    hasher.Write(buffer, sizeof(buffer));
+
+    // High-precision timestamp.
+    //
+    // Note that we also commit to a timestamp in the Fast seeder, so we indirectly commit to a
+    // benchmark of all the entropy gathering sources in this function).
+    SeedTimestamp(hasher);
 }
 
-void GetStrongRandBytes(unsigned char* out, int num)
+static void SeedSleep(CSHA512& hasher)
 {
-    assert(num <= 32);
-    CSHA512 hasher;
-    unsigned char buf[64];
+    // Everything that the 'fast' seeder includes
+    SeedFast(hasher);
 
-    // First source: OpenSSL's RNG
-    RandAddSeedPerfmon();
-    GetRandBytes(buf, 32);
-    hasher.Write(buf, 32);
+    // High-precision timestamp
+    SeedTimestamp(hasher);
 
-    // Second source: OS RNG
-    GetOSRand(buf);
-    hasher.Write(buf, 32);
+    // Sleep for 1ms
+    UninterruptibleSleep(std::chrono::milliseconds{1});
 
-    // Third source: HW RNG, if available.
-    if (GetHWRand(buf)) {
-        hasher.Write(buf, 32);
-    }
+    // High-precision timestamp after sleeping (as we commit to both the time before and after, this measures the delay)
+    SeedTimestamp(hasher);
 
-    // Combine with and update state
-    {
-        WAIT_LOCK(cs_rng_state, lock);
-        hasher.Write(rng_state, sizeof(rng_state));
-        hasher.Write((const unsigned char*)&rng_counter, sizeof(rng_counter));
-        ++rng_counter;
-        hasher.Finalize(buf);
-        memcpy(rng_state, buf + 32, 32);
-    }
```

Comment:
@UdjinM6 noting our difference with upstream

Reply:
Do you mean some other fixes are to be backported? This specific block applied cleanly on cherry-picking as far as I remember, no conflicts in these lines.

Reply:
@PastaPastaPasta can you clarify pls? I don't see any

Reply:
Upstream had https://github.com/bitcoin/bitcoin/pull/14955/files#diff-433f996b8149d6d6831ab1ee2978d4cc13dead5721c824175db806e4284ab15bL325-L328 instead of this, and I was just pointing out the difference

Reply:
https://github.com/dashpay/dash/pull/4411/files/4cbecd4510a62910519725aee8633d8b0166546e#diff-433f996b8149d6d6831ab1ee2978d4cc13dead5721c824175db806e4284ab15bL331-L334 👀  there is some confusion caused by diffs being displayed differently probably cause the code is the same as far as I can tell

Reply:
Sounds good, just wanted to point it out to make sure we weren't missing smth